backend/migrations/versions/2023_08_08_0946-18b1066d92c3_add_store_key_to_artefact_source.py

```diff
@@ -0,0 +1,103 @@
+"""Add store key to artefact source
+
+Revision ID: 18b1066d92c3
+Revises: 451ec6d4c102
+Create Date: 2023-08-08 09:46:34.620860+00:00
+
+"""
+from alembic import op
+from sqlalchemy.sql import text
+
+
+# revision identifiers, used by Alembic.
+revision = "18b1066d92c3"
+down_revision = "451ec6d4c102"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    conn = op.get_bind()
+
+    # Ensure all 'snap' artefacts have the "store" key
+    artefacts = conn.execute(
+        text(
+            """
+        SELECT a.id, a.source
+        FROM artefact AS a
+        JOIN stage AS s ON a.stage_id = s.id
+        JOIN family AS f ON s.family_id = f.id
+        WHERE f.name = 'snap'
+    """
+        )
+    ).fetchall()
+
+    for artefact in artefacts:
+        source_data = artefact.source
+        if "store" not in source_data:
+            source_data["store"] = "ubuntu"
+            conn.execute(
+                text(
+                    "UPDATE artefact SET source = :source WHERE id = :id",
+                ),
+                {"source": source_data, "id": artefact.id},
+            )
+    # Create the trigger function
+    op.execute(
+        """
+        CREATE OR REPLACE FUNCTION ensure_store_key_for_snap() RETURNS TRIGGER AS $$
+        BEGIN
+            IF (EXISTS (
+                SELECT 1
+                FROM stage s
+                JOIN family f ON s.family_id = f.id
+                WHERE s.id = NEW.stage_id AND f.name = 'snap'
+            ) AND NOT NEW.source ? 'store') THEN
+                RAISE EXCEPTION
+            'The "store" key is required in source for artefacts with the snap family';
+            END IF;
+
+            RETURN NEW;
+        END;
+        $$ LANGUAGE plpgsql;
+    """
+    )
+
+    # Attach the trigger to the artefact table
+    op.execute(
+        """
+        CREATE TRIGGER trigger_ensure_store_key_for_snap
+        BEFORE INSERT OR UPDATE
+        ON artefact
+        FOR EACH ROW
+        EXECUTE FUNCTION ensure_store_key_for_snap();
+    """
+    )
```

Comment:
I've tried multiple approaches and validation on the ORM level didn't work. I've tried all the options specified here: https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html

The problem is that on the ORM level, we can validate this field only when it's manually updated. So, it doesn't work when the function "get_or_create" is called since we use "insert" there. So, ORM approaches work half of the time.

Reply:
Oh great point. It didn't occur to me that using insert would bypass ORM. I remember facing this issue in a previous project. There we ended up deciding not to bypass ORM. But these findings of yours are great, now we can investigate more clearly our options.

The way I see it we have two questions:

1. Do we want to validate source field? (series, pocket, track, store, etc)
We need to add validations so it's a bit of work. But we catch errors early which is great.

2. Should we allow bypassing ORM? (running SQL without ORM)
If we bypass ORM we have more flexibility with SQL. But we'll lose all ORM functionality (checks, defaults, etc)

For question 1, I suggest to answer Yes (but we could also go with No). In that case I prefer if we add all source validation together in a separate PR.

For question 2 I suggest No. Having ORM then bypassing it can cause issues (as you've seen). Luckily you found the issues this time, but it could happen that we miss them the next time. So instead we'd have to change `get_or_create` to use ORM, by trying to create the object, and if it fails due to a uniqueness constraint, get it instead. We could also make this change part of the same separate PR as above.

wdyt?

Reply:
I can rename the function as "check_source_field" and we can extend it in another PR

Reply:
However, if we decide the answer for 2 should be No. Then we would place such validation in ORM no?

Reply:
Validation in ORM doesn't work as we want it.
If you use the approach with validation method, they "are only called by normal userland code; they are not issued when the ORM is populating the object".
If we use the approach with setters, we should create a private field `_source` and it will be called this way in the DB